qemu/tests/virtio_console.py

```diff
@@ -570,12 +571,154 @@ def test_loopback():
+                                      'id' in qdev.qtree) and (
+                                      qdev.qtree['id'] == qtree_check_port):
+                            return qbus.qtree['id']
+            return None
+        except AttributeError, err:
+            logging.warn("Monitor deson't supoort qtree skip this test, error"
+                         " log: %s " % err)
+
     @error_context.context_aware
     def test_interrupted_transfer():
         """
         This test creates loopback between 2 ports and interrupts transfer
         eg. by stopping the machine or by unplugging of the port.
         """
+        def _replug_loop():
+            def _port_unplug(port_idx):
+                dev = ports[port_idx]
+                portdev = get_virtio_port_by_name(vm, dev.name)
+                if not portdev:
+                    test.error("No port named %s" % dev.qemu_id)
+                if portdev.get_param("bus") == "None":
+                    bus_info = portdev.get_param("bus")
+                else:
+                    bus_info = get_bus_of_port(vm, portdev)
+                port_property = dict(id=portdev.get_param("id"),
+                                     name=portdev.get_param("name"),
+                                     chardev=portdev.get_param("chardev"),
+                                     bus=bus_info)
+                if not port_property["bus"]:
+                    test.error("Can't find the bus %s for this port" % bus_info)
+                (out, ver_out) = vm.devices.simple_unplug(portdev, vm.monitor)
+                if not ver_out:
+                    test.error("Error occured when unplug %s" % dev.name)
+                return port_property
+
+            def _port_plug(device, property):
+                portdev = qdevices.QDevice(device)
+                for key, value in {'id': property['id'],
+                                   'chardev': property['chardev'],
+                                   'name': property['name'],
+                                   'bus': property['bus']}.items():
+                    portdev.set_param(key, value)
+                (out, ver_out) = vm.devices.simple_hotplug(portdev, vm.monitor)
+                if not ver_out:
+                    test.error("Error occured when plug port %s." % property['name'])
+
+            def _pci_unplug(bus):
+                device = get_device_by_id(vm, str(bus).split('.')[0])
+                if not device:
+                    test.error("No bus %s in vm." % bus)
+                bus_property = dict(id=device.get_param("id"))
+                err = ""
+                out = device.unplug(vm.monitor)
+                if out:
+                    err += "\nHotplug monitor output: " + out
+                time.sleep(2)
+                ver_out = device.verify_unplug(out, vm.monitor)
+                deathline = time.time() + 5
+                while time.time() < deathline:
```

Comment:
In this case, you may want to use 'utils_misc.wait_for()'

Reply:
Thanks. The reason for this code block is that we didn't use qmp monitor's result to verify the result of unplugging, but check qtree's info everytime, and both `device.verify_unplug` and  `device.unplug` are non-blocking for the result, so I think the `utils_misc.wait_for()` can not be used this time.